File: C++/source/Framework/Messaging/LightHouse/LightPulse.cpp
/***************************************************************************
 *   Copyright (C) 20016 Vidyasagara Guntaka                               *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Messaging/LightHouse/LightPulse.h"
#include "Framework/Attributes/Attributes.h"
#include "Framework/Attributes/AttributeUUID.h"

namespace WaveNs
{

LightPulse::LightPulse (const string &name)
    : m_name (name),
      m_senderLightHousePort (0)
{
}

LightPulse::~LightPulse ()
{
}

string LightPulse::getName () const
{
    return m_name;
}

void LightPulse::setName (const string& name)
{
    m_name = name;
}

string LightPulse::getSenderIpAddressFromClientPerspective () const
{
    return m_senderIpAddressFromClientPerspective;
}

void LightPulse::setSenderIpAddressFromClientPerspective (const string &senderIpAddressFromClientPerspective)
{
    m_senderIpAddressFromClientPerspective = senderIpAddressFromClientPerspective;
}

Uuid LightPulse::getSenderLightHouseId () const
{
    return m_senderLightHouseId;
}

void LightPulse::setSenderLightHouseId (const Uuid &senderLightHouseId)
{
    m_senderLightHouseId = senderLightHouseId;
}

string LightPulse::getSenderLightHouseIpAddress () const
{
    return m_senderLightHouseIpAddress;
}

void LightPulse::setSenderLightHouseIpAddress (const string &senderLightHouseIpAddress)
{
    m_senderLightHouseIpAddress = senderLightHouseIpAddress;
}

SI32 LightPulse::getSenderLightHousePort () const
{
    return m_senderLightHousePort;
}

void LightPulse::setSenderLightHousePort (const SI32 &senderLightHousePort)
{
    m_senderLightHousePort = senderLightHousePort;
}

void LightPulse::setupAttributesForSerialization ()
{
    SerializableObject::setupAttributesForSerialization ();

    addSerializableAttribute (new AttributeString (&m_name,                                 "name"));
    addSerializableAttribute (new AttributeUUID   (&m_senderLightHouseId,                   "senderLightHouseId"));
    addSerializableAttribute (new AttributeString (&m_senderLightHouseIpAddress,            "senderLightHouseIpAddress"));
    addSerializableAttribute (new AttributeSI32   (&m_senderLightHousePort,                 "senderLightHousePort"));
    addSerializableAttribute (new AttributeString (&m_senderIpAddressFromClientPerspective, "senderIpAddressFromClientPerspective"));
}

string LightPulse::getLightPulseName (const char *pSerializedData)
{
    static const char *pAttributeStringAtBegining = "<Aname>";
    static const char *pAttributeStringAtEnding   = "</Aname>";
    static       char  sotString[256];

    const char *pBegin = strstr (pSerializedData, pAttributeStringAtBegining);
    const char *pValue = pBegin + strlen (pAttributeStringAtBegining);
    const char *pEnd   = strstr (pValue, pAttributeStringAtEnding);

    UI32 valueSize = pEnd - pValue;

    strncpy (sotString, pValue, valueSize);

    sotString[valueSize] = '\0';

    return (sotString);
}

}

File: Java/source/com/CxWave/Wave/Framework/Messaging/LightHouse/LightPulse.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Messaging.LightHouse;

import com.CxWave.Wave.Framework.ObjectModel.SerializableObject;
import com.CxWave.Wave.Framework.Type.SI32;
import com.CxWave.Wave.Framework.Type.Uuid;
import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Framework.Utils.String.WaveStringUtils;
import com.CxWave.Wave.Framework.Utils.Trace.WaveTraceUtils;

public class LightPulse extends SerializableObject
{
    private String m_name;
    private Uuid   m_senderLightHouseId                   = new Uuid ();
    private String m_senderLightHouseIpAddress            = "";
    private SI32   m_senderLightHousePort                 = new SI32 (0);
    private String m_senderIpAddressFromClientPerspective = "";

    public LightPulse (final String name)
    {
        m_name = name;
    }

    public String getName ()
    {
        return m_name;
    }

    public void setName (final String name)
    {
        m_name = name;
    }

    public Uuid getSenderLightHouseId ()
    {
        return m_senderLightHouseId;
    }

    public void setSenderLightHouseId (final Uuid senderLightHouseId)
    {
        m_senderLightHouseId = senderLightHouseId;
    }

    public String getSenderLightHouseIpAddress ()
    {
        return m_senderLightHouseIpAddress;
    }

    public void setSenderLightHouseIpAddress (final String senderLightHouseIpAddress)
    {
        m_senderLightHouseIpAddress = senderLightHouseIpAddress;
    }

    public SI32 getSenderLightHousePort ()
    {
        return m_senderLightHousePort;
    }

    public void setSenderLightHousePort (final SI32 senderLightHousePort)
    {
        m_senderLightHousePort = senderLightHousePort;
    }

    public String getSenderIpAddressFromClientPerspective ()
    {
        return m_senderIpAddressFromClientPerspective;
    }

    public void setSenderIpAddressFromClientPerspective (final String senderIpAddressFromClientPerspective)
    {
        m_senderIpAddressFromClientPerspective = senderIpAddressFromClientPerspective;
    }

    public static String getLightPulseNameForLightPulseClass (final Class<?> lightPulseClass)
    {
        WaveAssertUtils.waveAssert (null != lightPulseClass);

        Object object = null;

        try
        {
            object = lightPulseClass.newInstance ();
        }
        catch (final Exception exception)
        {
            WaveTraceUtils.fatalTracePrintf ("WaveEvent.getLightPulseNameForLightPulseClass : %s message class could not be instantiated.  Details : %s", lightPulseClass.getName (), exception.toString ());

            WaveAssertUtils.waveAssert ();
        }

        final LightPulse lightPulse = (LightPulse) object;

        WaveAssertUtils.waveAssert (null != lightPulse);

        final String lightPulseName = lightPulse.getName ();

        WaveAssertUtils.waveAssert (WaveStringUtils.isNotBlank (lightPulseName));

        return (lightPulseName);
    }

    public static String getLightPulseName (final String serializedLightPulseString)
    {
        if (WaveStringUtils.isBlank (serializedLightPulseString))
        {
            return ("");
        }

        final int seralizedLightPulseStringLength = serializedLightPulseString.length ();
        final String attributeNameLocatorString = "4#name";
        final int attributeNameLocatorStringLength = attributeNameLocatorString.length ();
        final int attributeNameLocatorStringIndex = serializedLightPulseString.indexOf (attributeNameLocatorString, 0);

        if (-1 == attributeNameLocatorStringIndex)
        {
            return ("");
        }

        final int attributeValueLengthStartIndex = attributeNameLocatorStringIndex + attributeNameLocatorStringLength;

        WaveAssertUtils.waveAssert (attributeValueLengthStartIndex < seralizedLightPulseStringLength);

        final int attributeValueLengthEndMarkerIndex = serializedLightPulseString.indexOf ("#", attributeValueLengthStartIndex);

        if (-1 == attributeValueLengthEndMarkerIndex)
        {
            return ("");
        }

        final int attributeValueLength = Integer.valueOf (serializedLightPulseString.substring (attributeValueLengthStartIndex, attributeValueLengthEndMarkerIndex));

        final int attributeValueStartIndex = attributeValueLengthEndMarkerIndex + 1;
        final int attributeValueEndIndex = attributeValueStartIndex + attributeValueLength;

        WaveAssertUtils.waveAssert (attributeValueStartIndex < seralizedLightPulseStringLength);
        WaveAssertUtils.waveAssert (attributeValueEndIndex < seralizedLightPulseStringLength);

        final String attributeValue = serializedLightPulseString.substring (attributeValueStartIndex, attributeValueEndIndex);

        return (attributeValue);
    }
}
